### PR TITLE
[Bug] flink cluster management bug fixed

### PR DESCRIPTION
**Pull request by wolfboys**
[Bug] flink cluster management bug fixed

## What changes were proposed in this pull request

1). add remote-cluster connection  check
2) add yarn-session-cluster connection  check
3) simplify add cluster parameter settings
4) cluster run status bug fixed

Issue Number: close #xxx 

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): no

### CODE REVIEW

**Comment by wolfboys** [comment]
cc @macksonmu @xujiangfeng001 PTAL thanks

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/mapper/ApplicationMapper.java [hunk 1]
```diff
@@ -57,4 +57,6 @@ public interface ApplicationMapper extends BaseMapper<Application> {
     Boolean existsByJobName(@Param("jobName") String jobName);
 
     List<Application> getByProjectId(@Param("projectId") Long id);
+
+    boolean existsRunningJobByClusterId(@Param("clusterId")Long clusterId);
```
Comment (muchunjin) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/mapper/ApplicationMapper.java:61]: missing space before `Long`

File: streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml [hunk 2]
```diff
@@ -98,6 +99,14 @@
         limit 1
     </select>
 
+    <select id="existsRunningJobByClusterId" resultType="java.lang.Boolean" parameterType="java.lang.Long">
+        select count(1)
+        from t_flink_app
+        where flink_cluster_id = #{clusterId}
```
Comment (muchunjin) [inline streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml:112]: I didn't understand, what is the difference between `cluster_id` and `flink_cluster_id`

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java [hunk 3]
```diff
@@ -138,23 +144,34 @@ public URI getActiveAddress() {
...
+            if (address == null) {
                 return false;
             }
-            try {
-                HttpClientUtils.httpGetRequest(url, RequestConfig.custom().setConnectTimeout(2000).build());
-                return true;
-            } catch (Exception ignored) {
-                //
+            String[] array = address.split(",");
+            for (String url : array) {
```
Comment (muchunjin) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: Why is there a cycle here?

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java [hunk 4]
```diff
@@ -519,6 +529,21 @@ public boolean existsByTeamId(Long teamId) {
         return baseMapper.existsByTeamId(teamId);
     }
 
+    @Override
+    public boolean existsRunningJobByClusterId(Long clusterId) {
+        boolean exists = baseMapper.existsRunningJobByClusterId(clusterId);
+        if (!exists) {
+            for (Application application:FlinkTrackingTask.getAllTrackingApp().values()) {
```
Comment (muchunjin) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: no spaces around the colon

File: streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml (hunk 2 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml:112]: > I didn't understand, what is the difference between `cluster_id` and `flink_cluster_id`

cluster_id: Such as the applicationId of the yarn-session and the k8s cluster Id
flink_cluster_id: Clusters managed by streampark, from t_flink_cluster id field

File: streampark-console/streampark-console-webapp/src/locales/lang/en/flink/app.ts [hunk 5]
```diff
@@ -41,7 +41,6 @@ export default {
   podTemplate: 'Kubernetes Pod Template',
   flinkCluster: 'Flink Cluster',
   yarnQueue: 'Yarn Queue',
-  yarnSessionClusterId: 'Yarn Session ClusterId',
```
Comment (muchunjin) [inline streampark-console/streampark-console-webapp/src/locales/lang/en/flink/app.ts:44]: When adding an existing yarn session cluster, many people may not know what `address` and `Yarn Session Cluster` are. It is recommended to give a hint, for example, `address` is the jobmanager address, such as http://hh-hx199w:35154, `Yarn Session Cluster` Is a value similar to application_1656423003647_19237

File: streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml (hunk 2 above)
Comment (muchunjin) [reply streampark-console/streampark-console-service/src/main/resources/mapper/core/ApplicationMapper.xml:112]: ok i got it

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 3 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > Why is there a cycle here?

For example,   the address may be a set of addresses, such as: 192.168.0.1, 192.168.0.2, 192.168.0.3

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java (hunk 4 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: > no spaces around the colon

Thanks for your review, I will improve it later

**Comment by 1996fanrui** [comment]
> cluster run status bug fixed

What's the bug? And please update the PR and commit message. It's better to write the bug info. Developer, reviewer and other contributor will be easy to know what the PR do.

**Review by muchunjin — COMMENTED** [review]
LGTM. Code test passed, thanks for your hard work,

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java [hunk 6]
```diff
@@ -138,23 +144,34 @@ public URI getActiveAddress() {
...
+                        restUrl = url + "/overview";
+                    } else {
+                        restUrl = url + "/proxy/" + this.clusterId + "/overview";
+                    }
+                    String result = HttpClientUtils.httpGetRequest(restUrl, RequestConfig.custom().setConnectTimeout(2000).build());
+                    JacksonUtils.read(result, Overview.class);
+                    return true;
+                } catch (Exception ignored) {
+                    //
+                }
```
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: Can 2 `try ... catch` be simplified to one?

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java [hunk 7]
```diff
@@ -138,23 +144,34 @@ public URI getActiveAddress() {
...
             }
+            return false;
         }
         return false;
```
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: Can 3 `return false` be simplified to one?

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 3 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: Same question: Isn't the address JM url? 

If the address is JM url, why is it a list?

Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > Same question: Isn't the address JM url?
> 
> If the address is JM url, why is it a list?

e.g:
1): cluster is  flink standalone cluster, is a set of addresses
2): cluster is yarn-session cluster, is yarn HA addresses

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 7 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > 

The first `try..catch` is to ensure that a group of addresses, each is a legitimate url, if any of them is an invalid url then return,
The second `try...catch` is to find an active and connected address in a group of addresses

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 3 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: As I understand, yarn session cluster is a yarn job. It just has a JobManager and when old JM is crashed, yarn will start a new JM, right?

If so, I don't understand the HA address.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 7 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: From your reply, I think they can be simplified . I still don't understand why don't simplify them.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java [hunk 8]
```diff
@@ -519,6 +529,26 @@ public boolean existsByTeamId(Long teamId) {
...
+            for (Application application : FlinkTrackingTask.getAllTrackingApp().values()) {
+                if (clusterId.equals(application.getFlinkClusterId())) {
+                    if (FlinkAppState.RUNNING.equals(application.getFlinkAppStateEnum())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return exists;
+    }
```
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: This is bad case about code style. You can take a look Flink Code style in your free time, it's useful for  high quality code of StreamPark.

https://flink.apache.org/contributing/code-style-and-quality-common.html

image

image

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java [hunk 9]
```diff
@@ -164,11 +181,14 @@ public Map<String, String> getFlinkConfig() throws MalformedURLException, JsonPr
...
-        confList.forEach(k -> config.put(k.get("key"), k.get("value")));
-        return config;
+        if (StringUtils.isNotEmpty(json)) {
+            List<Map<String, String>> confList = JacksonUtils.read(json, new TypeReference<List<Map<String, String>>>() {
+            });
+            Map<String, String> config = new HashMap<>(0);
+            confList.forEach(k -> config.put(k.get("key"), k.get("value")));
+            return config;
+        }
+        return Collections.emptyMap();
```
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]:
```suggestion
        if (StringUtils.isEmpty(json)) {
            return Collections.emptyMap();
        }
        
        List<Map<String, String>> confList = JacksonUtils.read(json, new TypeReference<List<Map<String, String>>>() {
            });
        Map<String, String> config = new HashMap<>(0);
        confList.forEach(k -> config.put(k.get("key"), k.get("value")));
        return config;
```

**Comment by 1996fanrui** [comment]
> > cluster run status bug fixed
> 
> What's the bug? And please update the PR and commit message. It's better to write the bug info. Developer, reviewer and other contributor will be easy to know what the PR do.

Hi @wolfboys , please help take a look, thanks~

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 7 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > From your reply, I think they can be simplified . I still don't understand why don't simplify them.

   e.g: address is : http://192.168.0.1,xxyy:a.b.c,192.168.1.2

  ```
    String[] array = address.split(",");
            for (String url : array) {
                try {
                    new URI(url);
                } catch (Exception ignored) {
                    return false;
                }
                try {
                    String restUrl;
                    if (ExecutionMode.REMOTE.equals(this.getExecutionModeEnum())) {
                        restUrl = url + "/overview";
                    } else {
                        restUrl = url + "/proxy/" + this.clusterId + "/overview";
                    }
                    String result = HttpClientUtils.httpGetRequest(restUrl, RequestConfig.custom().setConnectTimeout(2000).build());
                    JacksonUtils.read(result, Overview.class);
                    return true;
                } catch (Exception ignored) {
                    //
                }
            }
```

then parse `xxyy:a.b.c` first try..catch will return, second will find only one if connection is ok and return.

Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > > From your reply, I think they can be simplified . I still don't understand why don't simplify them.
> 
> e.g: address is : http://192.168.0.1,xxyy:a.b.c,192.168.1.2
> 
> ```
>   String[] array = address.split(",");
>           for (String url : array) {
>               try {
>                   new URI(url);
>               } catch (Exception ignored) {
>                   return false;
>               }
>               try {
>                   String restUrl;
>                   if (ExecutionMode.REMOTE.equals(this.getExecutionModeEnum())) {
>                       restUrl = url + "/overview";
>                   } else {
>                       restUrl = url + "/proxy/" + this.clusterId + "/overview";
>                   }
>                   String result = HttpClientUtils.httpGetRequest(restUrl, RequestConfig.custom().setConnectTimeout(2000).build());
>                   JacksonUtils.read(result, Overview.class);
>                   return true;
>               } catch (Exception ignored) {
>                   //
>               }
>           }
> ```
> 
> then parse `xxyy:a.b.c` first try..catch will return, second will find only one if connection is ok and return.

first try...catch Will check whether each url is ok, but the second is not

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java (hunk 8 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: > This is bad case about code style. You can take a look Flink Code style in your free time, it's useful for high quality code of StreamPark.
> 
> https://flink.apache.org/contributing/code-style-and-quality-common.html
> 
> image
> 
> image

hi fanrui:

Thanks for your feedback. If you read the code carefully, you will find that the situation is different from what you describe

the following code is improved:

```
@Override
    public boolean existsRunningJobByClusterId(Long clusterId) {
        boolean exists = baseMapper.existsRunningJobByClusterId(clusterId);
        if (!exists) {
            for (Application application : FlinkTrackingTask.getAllTrackingApp().values()) {

                if (clusterId.equals(application.getFlinkClusterId()) && FlinkAppState.RUNNING.equals(application.getFlinkAppStateEnum()) ) {
                        return true;
                }
            }
        }
        return exists;
    }
```

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 7 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: After read this code 3 times, I think I understand.

The first `try catch` will return directly when any url isn't legal, right?  For example, the url isn't legal, the next url won't be check.

If right, I think the code should be refactored. It's very very hard to read, I read it too many times to understand it. 😭

How about this? The `isLegalUrl` is an util method.
```
for (String url : array) {
  if(!isLegalUrl(url)) {
    return false;
  }
  // try to connect yarn ...
}
```

If It's still wrong, please correct me.

**Comment by wolfboys** [comment]
> 

1). add remote-cluster connection check
2) add yarn-session-cluster connection check
3) simplify add cluster parameter settings
4) cluster run status bug fixed

issue: #2091, #2092

**Comment by wolfboys** [comment]
> 



> > > cluster run status bug fixed
> > 
> > 
> > What's the bug? And please update the PR and commit message. It's better to write the bug info. Developer, reviewer and other contributor will be easy to know what the PR do.
> 
> Hi @wolfboys , please help take a look, thanks~

1). add remote-cluster connection check
2) add yarn-session-cluster connection check
3) simplify add cluster parameter settings
4) cluster run status bug fixed

issue: https://github.com/apache/incubator-streampark/issues/2091, https://github.com/apache/incubator-streampark/issues/2092

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java (hunk 8 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: How about this?

```
    @Override
    public boolean existsRunningJobByClusterId(Long clusterId) {
        boolean exists = baseMapper.existsRunningJobByClusterId(clusterId);
        if (exists) {
            return true;
        }

        for (Application application : FlinkTrackingTask.getAllTrackingApp().values()) {
            if (clusterId.equals(application.getFlinkClusterId()) && FlinkAppState.RUNNING.equals(application.getFlinkAppStateEnum()) ) {
                return true;
            }
        }
        return false;
    }
```

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 3 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > As I understand, yarn session cluster is a yarn job. It just has a JobManager and when old JM is crashed, yarn will start a new JM, right?
> 
> If so, I don't understand the HA address.

I don't want to discuss this part too much, If you set a single url here, corresponding to a unit of yarn node, once the node hangs, the subsequent can not do other work.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 7 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: > After read this code 3 times, I think I understand.
> 
> The first `try catch` will return directly when any url isn't legal, right? For example, the url isn't legal, the next url won't be check.
> 
> If right, I think the code should be refactored. It's very very hard to read, I read it too many times to understand it. 😭
> 
> How about this? The `isLegalUrl` is an util method.
> 
> ```
> for (String url : array) {
>   if(!isLegalUrl(url)) {
>     return false;
>   }
>   // try to connect yarn ...
> }
> ```
> 
> If It's still wrong, please correct me.

ok, good. 👍

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java (hunk 3 above)
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/FlinkCluster.java]: >  It just has a JobManager and when old JM is crashed, yarn will start a new JM, right?

As I understand, yarn will start a new JM after old JM crashed. So I wonder how to get the HA address. Maybe I missed some background.

**Comment by wolfboys** [comment]
cc @1996fanrui:

Thanks for your review and great comments, the latest code has been submitted. Please review again when you have time, thanks very much

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java (hunk 8 above)
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: > How about this?
> 
> ```
>     @Override
>     public boolean existsRunningJobByClusterId(Long clusterId) {
>         boolean exists = baseMapper.existsRunningJobByClusterId(clusterId);
>         if (exists) {
>             return true;
>         }
> 
>         for (Application application : FlinkTrackingTask.getAllTrackingApp().values()) {
>             if (clusterId.equals(application.getFlinkClusterId()) && FlinkAppState.RUNNING.equals(application.getFlinkAppStateEnum()) ) {
>                 return true;
>             }
>         }
>         return false;
>     }
> ```

I also submitted, it seems that the code is exactly the same as this 😅

**Comment by MonsterChenzhuo** [comment]
LGTM

**Review by MonsterChenzhuo — APPROVED** [review]

**Review by lvshaokang — APPROVED** [review]
LGTM

**Pull request merged**